grr/server/grr_response_server/output_plugins/elasticsearch_plugin.py

```diff
@@ -0,0 +1,155 @@
+          "installation, e.g. 'https://myelasticsearch.example.com:9200'.")
+
+    self._url = urlparse.urljoin(url, BULK_OPERATIONS_PATH)
+
+  def ProcessResponses(self, state: rdf_protodict.AttributedDict,
+                       responses: List[rdf_flows.GrrMessage]) -> None:
+    """See base class."""
+    client_id = self._GetClientId(responses)
+    flow_id = self._GetFlowId(responses)
+
+    client = self._GetClientMetadata(client_id)
+    flow = self._GetFlowMetadata(client_id, flow_id)
+
+    events = [self._MakeEvent(response, client, flow) for response in responses]
+    self._SendEvents(events)
+
+  def _GetClientId(self, responses: List[rdf_flows.GrrMessage]) -> Text:
+    client_ids = {msg.source.Basename() for msg in responses}
+    if len(client_ids) > 1:
+      raise AssertionError((
+          "ProcessResponses received messages from different Clients {}, which "
+          "violates OutputPlugin constraints.").format(client_ids))
+    return client_ids.pop()
+
+  def _GetFlowId(self, responses: List[rdf_flows.GrrMessage]) -> Text:
+    flow_ids = {msg.session_id.Basename() for msg in responses}
+    if len(flow_ids) > 1:
+      raise AssertionError(
+          ("ProcessResponses received messages from different Flows {}, which "
+           "violates OutputPlugin constraints.").format(flow_ids))
+    return flow_ids.pop()
+
+  def _GetClientMetadata(self, client_id: Text) -> export.ExportedMetadata:
+    info = data_store.REL_DB.ReadClientFullInfo(client_id)
+    metadata = export.GetMetadata(client_id, info)
+    metadata.timestamp = None  # timestamp is sent outside of metadata.
+    return metadata
+
+  def _GetFlowMetadata(self, client_id: Text,
+                       flow_id: Text) -> api_flow.ApiFlow:
+    flow_obj = data_store.REL_DB.ReadFlowObject(client_id, flow_id)
+    return api_flow.ApiFlow().InitFromFlowObject(flow_obj)
+
+  def _MakeEvent(self, message: rdf_flows.GrrMessage,
+                 client: export.ExportedMetadata,
+                 flow: api_flow.ApiFlow) -> JsonDict:
+
+    event = {
+      "client": _ToDict(client),
+      "flow": _ToDict(flow),
+      "resultType": message.args_rdf_name,
+      "result": _ToDict(message.payload),
+    }
+
+    if self.args.tags:
+      event["tags"] = list(self.args.tags)
+
+    return event
+
+  def _SendEvents(self, events: List[JsonDict]) -> None:
+    """Uses the Elasticsearch bulk API to index all the events in a single
+    request.
+    https://www.elastic.co/guide/en/elasticsearch/reference/7.1/docs-bulk.html
+    """
+    if self._token:
+      headers = {"Authorization": "Basic {}".format(self._token)}
+    else:
+      headers = {}
+
+    index_command = json.Dump({"index": {"_index": self._index}}, indent=None)
+
+    # Each index operation is two lines, the first defining the index settings,
+    # the second is the actual document to be indexed
+    data = "\n".join(
+      index_command + "\n" + json.Dump(event, indent=None)
```

Comment:
Please use string formatting (f-strings or `format()`) instead of manual string concatenation, e.g.:

```py
    data = "\n".join(
      f"{index_command}\n{json.Dump(event, indent=None)}"
      for event in events)
```

Reply:
Done - new code is:

```
data = "\n".join([
      "{}\n{}".format(index_command, json.Dump(event, indent=None))
      for event in events])
```